### ISSUE TITLE
Transactional Tests

### ISSUE THREAD

**Issue by cjancsar** [sequelize/sequelize#11408]
### Is your feature request related to a problem? Please describe.
It is extremely common to want to run tests in an isolated way (regardless of the testing framework used). Sequelize should have a way to wrap test cases in transactions so that after each test is transactions are rolled back so DB state is not changed. This is a very common pattern in more mature solutions (i.e Django ORM).

Is there any modern update to this pattern described here: #4823

### Describe the solution you'd like
jest example"
```js
describe('some test', () => {
    it('Does a thing with DB', done => {
         ... some function that changes DB state
         done()
    })
})
```

After the above test is run (success / fail) the database should be back in it original state.

### Why should this be in Sequelize
It is extremely common to want to run tests in an isolated way (regardless of the testing framework used)

### Describe alternatives/workarounds you've considered
Currently you can truncate the tables that you are working on (with cascade if required) to empty the tables, although this can effect test performance.

### Additional context
Reference for how Django does it: https://docs.djangoproject.com/en/2.2/topics/testing/tools/#django.test.TestCase

### Is this issue dialect-specific?

- [ X ] No. This issue is relevant to Sequelize as a whole.

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [ X ] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by leifb** [comment]
@cjancsar did you get anywhere with this? I am trying to do the same, but it looks like it is not currently working.

Here is my current approach (which is not working):

1. Init Sequelize with cls-hooks: `Sequelize.useCLS(my_namespace)`
2. Use [mocha root-level hooks](https://mochajs.org/#root-level-hooks) to
    - create a transaction: `wrapperTransaction = await sequelize.transaction()`
    - load fixtures using `sequelize-fixtures`
3. Use mocha's beforeAll to
    - create a per-test-transaction: `currentTestTransaction = await sequelize.transaction({ transaction: wrapperTransaction })` (Which is using undocumented functionality #10840)
4. Use mochas's afterAll to
    - Roll back changes done by the test: `await currentTestTransaction.rollback()`
5. Use root level hooks to
    - Roll back the fixtures: `await wrapperTransaction.rollback()`

It would be really neat if that worked, but my guess is that sequelize does not allow using unmanaged transactions outside of the `then()`-block, of where they where supposed to be used (as described in the [documentation](https://sequelize.org/master/manual/transactions.html)). @papb it would be nice if you could confirm that.

I am using a separate file to handle all the fixture and transaction stuff. The idea is that the tests are written as if they are not being run in a transaction. Or in other words, I want to add the transactions without having to change all my tests.

**Comment by cjancsar** [comment]
@leifb I just created a helper function to manually clear any related tables after each test.

**Comment by leifb** [comment]
@cjancsar Thanks for the reply. We did that before and the whole idea was to stop truncating and move to transactions in order to reduce the time our tests take. Before, our tests took about two to three minutes. Now using Transactions they are done in ~8 seconds.

We got around the limitations of sequelize by using a custom `it` method instead of mochas `it`. It's not a pretty solution, but it works for us. Here's how:

  - Initialize sequelize using a `cls-hooked` namespace.
  - Using mochas global hooks `before` and `after` to load fixtures (`sequelize-fixtures`) in the beginning and truncate after all tests are done.
  - Create a custom `it`, calls mochas `it` and starts a transaction which then calls the actual test.
  - import this new `it` in all of the test files you want to use transacitons.

Here is how our `it` looks:

```ts
module.exports.it = function (title, fn) {
    return global.it(title, function () {
        return new Promise((resolve, reject) => {
            db.sequelize.transaction(t => {
                return fn().then(() => {
                    resolve();
                }).catch(err => {
                    reject(err);
                }).finally(() => {
                    throw "Rollback";
                });
            }).catch(err => {
                if (err === "Rollback") return;
                reject(err);
                console.log("Error in test:");
                console.log(err);
            });
        });
    });
}
```

Basically what is happening is that every test gets wrapped around a transaction-call, which then calls the test. Therefore there is one limitation with this approach: We can only write asyc mocha tests instead of using the `done` method. I feel like getting around this would require a more sophisticated `it` that awaits the calling of the `done` method. I am not sure if it is currently possible. But rewriting a few tests was not a big deal for us.

**Comment by dkzlv** [comment]
@leifb I tried your solution but I seem to have problems with manual transactions. As far as I understand, queries that have manual transactions assigned will not be rolled back after the test — at least it's what I get.

This code
```typescript
async function doStuff(data) {
  const transaction = await sequelize.transaction()
  const res = await SomeModel.create(data, { transaction })
  await transaction.commit()
  return res
}
```
with this test
```typescript
describe('file', () => {
  // I'm using wrapped `it` from your example
  it('works', async () => {
    await doStuff({})
  })
})
```

will produce data that persists after test execution. Also in some strange cases (like 10% of clean runs) it will lock sqlite for some unknown reason.

Have you run into these problems? If so how did you solve them?

**Comment by leifb** [comment]
@dkzlv  As far as I know, I was not using any other transactions. I am also using postgresql, so maybe there is a bug specific to sqlite.

The solution I posted is still working fine for me.

**Comment by Motii1** [comment]
@dkzlv I had similar problem and I solved it by enabling `CLS` like this:
```ts
if (process.env.NODE_ENV === 'test') {
  const namespace = cls.createNamespace('namespace');
  Sequelize.useCLS(namespace);
}
```

**Comment by brian-petersen** [comment]
@dkzlv, I ran into the same issue. @Motii1, your solution didn't quite work as needed for me.

I had to use some undocumented features to get this to work as I expected:

```js
const sequelize = new Sequelize(params);

if (config.env === 'test') {
  const namespace = cls.createNamespace('test-transaction');
  Sequelize.useCLS(namespace);

  const originalTransactionFn = sequelize.transaction.bind(sequelize);
  sequelize.transaction = function (options, autoCallback) {
    const transaction = namespace.get('transaction');

    if (typeof options === 'function') {
      return originalTransactionFn({ transaction }, options);
    }

    return originalTransactionFn({...options, transaction }, autoCallback);
  }
}
```

**Comment by mat813** [comment]
Ok, so, I got this to work using Jest. `Logs` is one of my models, you can use whatever model you have, is is only used to get to the sequelize object to create a transaction (maybe there is a better way, but, it works)

Description in comments.

```ts
let transaction: Transaction | undefined;
let wrapperTransaction: Transaction | undefined;
let seq: ReturnType<typeof initSequelize>;

// Create namespace
const namespace = createNamespace('default-namespace');
// and assigns it to Sequelize
Sequelize.useCLS(namespace);

// Initializes Sequelize, create a first level transaction, a context, and enters this context.
const beforeAllBase = async () => {
  seq = init(); // calls my initSequelize()

  transaction = await seq.Logs.sequelize?.transaction({
    autocommit: false,
  });

  const context = namespace.createContext();
  namespace.enter(context);
  namespace.set('transaction', transaction);
};

// Wrap each test in a nested transaction
const beforeEachBase = async () => {
  wrapperTransaction = await seq.Logs.sequelize?.transaction({
    autocommit: false,
    transaction,
  });

  namespace.set('transaction', wrapperTransaction);
};

// Rollback each nested transaction
const afterEachBase = async () => {
  await wrapperTransaction?.rollback();
};

// Rollback the actual transaction
const afterAllBase = async () => {
  await transaction?.rollback();
};

export const setupHooks = () => {
  beforeAll(beforeAllBase);
  beforeEach(beforeEachBase);
  afterEach(afterEachBase);
  afterAll(afterAllBase);
};
```

**Comment by OmriSama** [comment]
Just as a heads up for anyone looking at this issue, Sequelize has deprecated `cls-hooked` in favor of using `AsyncLocalStorage` from Node, while also discouraging unmanaged transactions: https://github.com/sequelize/sequelize/pull/15292/

**Comment by ephys** [comment]
What you said is still true: It's still AsyncLocalStorage, we just kept the name CLS. Functionally, it should behave the same (except that it's enabled by default, and you don't need to install a third-party package). This means that in Sequelize 7, the only way to set the CLS transaction is by calling `sequelize.transaction()`

It's also true that we discourage using unmanaged transactions (the feature was moved to `sequelize.startUnmanagedTransaction()`), because it's easy to forget calling rollback when there is an error, and they don't interact with CLS

---

As for this issue, I would not recommend using transactions as a way of rolling back your test's changes as you're likely testing code that should start its own transaction. I would recommend using either `sequelize.sync({ force: true })`, `sequelize.destroyAll()` (new from Sequelize 7) or `sequelize.truncate`, depending on your needs

If you really want a transaction per test, this is the solution I would opt for: https://github.com/sequelize/sequelize/issues/11408#issuecomment-563962996
